File: 454.4-sum-ii.java
/*
 * @lc app=leetcode id=454 lang=java
 *
 * [454] 4Sum II
 */

// @lc code=start

import java.util.HashMap;
import java.util.Map;

class Solution {
    public int fourSumCount(int[] nums1, int[] nums2, int[] nums3, int[] nums4) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i : nums1) {
            for (int j : nums2) {
                int s = i + j;
                map.put(s, map.getOrDefault(s, 0) + 1);
            }
        }
        int cnt = 0;
        for (int i : nums3) {
            for (int j : nums4) {
                int s = i + j;
                cnt += map.getOrDefault(-s, 0);
            }
        }
        return cnt;
    }
}
// @lc code=end

File: 454.4-sum-ii.cpp
/*
 * @lc app=leetcode id=454 lang=cpp
 *
 * [454] 4Sum II
 */

// @lc code=start

#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int fourSumCount(vector<int>& nums1, vector<int>& nums2, vector<int>& nums3, vector<int>& nums4) {
        unordered_map<int, int> seen12, seen34;
        for(auto i = nums1.begin(); i != nums1.end(); ++i) {
            for (auto j = nums2.begin(); j != nums2.end(); ++j) {
                seen12[*i + *j] += 1;
            }
        }
        for(auto i = nums3.begin(); i != nums3.end(); ++i) {
            for (auto j = nums4.begin(); j != nums4.end(); ++j) {
                seen34[*i + *j] += 1;
            }
        }
        int ret = 0;
        for (auto i = seen12.begin(); i != seen12.end(); ++i) {
            auto j = seen34.find(-i->first);
            if (j != seen34.end()) {
                ret += i->second * j->second;
            }
        }
        return ret;
    }
};
// @lc code=end
